### PR TITLE
Task US012 Done

### PR DESCRIPTION
**Pull request by BartlomiejMilewski**
### 🛠 Changes being made

Here give examples of the changes you've made in this pull request. Include an itemized list if you can. It'll help the reviewer

### ✨ Definition of Done (all the points need to be filled out before merging)

- [ ] Meets Acceptance Criteria
- [ ] Tested
- [ ] Code Review made by team 2 members
- [ ] UX approved
- [ ] Product Owner Approved

### 🧪 Non-Functional Requirements

- [ ] Automated tests should cover acceptance criterias (Unit tests, Integration/API/Component tests)
- [ ] All input must be validated (client side & API)
- [ ] Sensitive data should not be stored in the URL
- [ ] No secrets in codesource

### CODE REVIEW

File: product-backend/pom.xml [hunk 1]
```diff
@@ -113,6 +113,11 @@
             <artifactId>byte-buddy</artifactId>
             <version>1.14.5</version>
         </dependency>
+        <dependency>
+            <groupId>io.swagger</groupId>
+            <artifactId>swagger-annotations</artifactId>
+            <version>1.6.0</version>
+        </dependency>
```
Comment (ljercha) [inline product-backend/pom.xml]: do you need it?
```suggestion
        <dependency>
            <groupId>io.swagger</groupId>
            <artifactId>swagger-annotations</artifactId>
            <version>1.6.0</version>
        </dependency>
```

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+@Tag(name = "Group A")
+@Path("/api")
+
+public class JobRoleController {
+    private final static Logger logger = LoggerFactory.getLogger(JobRoleService.class);
+    private final JobRoleService jobRoleService = new JobRoleService(new JobRoleDao(), new JobRoleValidator());
+
+    public JobRoleController() throws FailedToCreateNewJobRoleException {
+    }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java]: Constructor not used
```suggestion

```

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response createNewJobRole(JobRoleRequest jobRole) {
+        try {
+            return Response.ok(jobRoleService.createNewJobRole(jobRole)).build();
+        } catch (FailedToCreateNewJobRoleException e) {
+            logger.error("Failed to create new Job Role! Error: {}", e.getMessage());
+
+            return Response.serverError().entity(new ErrorResponse(e.getMessage())).build();
+        } catch (SQLException e) {
+            logger.error("Job Role already exists! Error : {}", e.getMessage());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java]: The exception is generic but the message is specific "Job Role already exists".

I you want to verify the job already exist you should reuse method from Adrian PR in JobRoleService
```
Optional<JobRole> jobRoleExist = jobRoleDao.getJobRoleById(jobRole.getId());
if (jobRoleExist.isPresent()) {
 throw new JobRoleAlreadyExistException();
}
```
Then in the controller catch it and return 400(with a message)

File: product-backend/src/main/java/org/kainos/ea/service/JobRoleValidator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        if (jobRole.getName().length() > 64) {
+            return "Name longer than 64 characters";
+        }
+
+        if (jobRole.getDescription().length() > 2000) {
+            return "Description longer than 2000 characters";
+        }
+
+        if (jobRole.getSharePointLink().length() > 2137) {
+            return "SharePointLink longer than 2137 characters";
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/JobRoleValidator.java]: Add a validation to verify starts with (https:// or http://)

File: product-backend/src/main/java/org/kainos/ea/service/JobRoleValidator.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        }
+
+        return null;
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/JobRoleValidator.java]: you could use optional here as well no avoid operating on nulls (can be very dangerous!)
```suggestion
public class JobRoleValidator {

    public Optional<String> isValidJobRole(JobRoleRequest jobRole) {

        if (jobRole.getName().length() > 64) {
            return Optional.of("Name longer than 64 characters");
        }

        if (jobRole.getDescription().length() > 2000) {
            return Optional.of("Description longer than 2000 characters");
        }

        if (jobRole.getSharePointLink().length() > 2137) {
            return Optional.of("SharePointLink longer than 2137 characters");
        }

        return Optional.empty();
```

File: product-web-app/tsconfig.json [hunk 6]
```diff
@@ -28,7 +28,7 @@
 
     "moduleResolution": "Node16",
     /* Modules */
-    "module": "ES2020",                                /* Specify what module code is generated. */
+    "module": "Node16",                                /* Specify what module code is generated. */
```
Comment (ljercha) [inline product-web-app/tsconfig.json]: You shouldn't need to change that

File: product-web-app/common/constants.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export const API_URL = `${process.env.API_URL}`;
+
+export const API = {
+  JOBROLES: '/api/admin/job-roles',
```
Comment (ljercha) [inline product-web-app/common/constants.ts]:
```suggestion
  JOB_ROLES: '/api/admin/job-roles',
```

File: product-web-app/controller/jobRoleController.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    private jobRoleService = new JobRoleService(new JobRoleValidator);
+
+    appRoutes(app: Application) {
+
+        app.get('/admin/add-job-roles', async (req: Request, res: Response) => {
+            res.render('add-new-job-role');
+        });
+
+        app.post('/admin/add-job-roles', async (req: Request, res: Response) => {
+            const data: JobRole = req.body;
```
Comment (ljercha) [inline product-web-app/controller/jobRoleController.ts]: Sanitize data

```suggestion
            const data: JobRole = req.body;
            data.name = sanitizeHtml(data.name).trim();
            data.description = sanitizeHtml(data.description).trim();
            data.sharePointLink = sanitizeHtml(data.sharePointLink).trim();
```

File: product-web-app/views/add-new-job-role.html [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            <div class="col-md-12">
+                <h1>Add Job Role</h1>
+            </div>
+        </div>
+    </header>
+    <main class="container mt-4">
+        <section id="form" class="bg-light py-5">
+            <div class="container"></div>
+            <p>Enter Job Role details</p>
+            {% if errormessage %}
```
Comment (ljercha) [inline product-web-app/views/add-new-job-role.html]:
```suggestion
            {% if errorMessage %}
```

**Review by ljercha — COMMENTED** [review]
Not so bad!

File: product-web-app/tests/jobRoleService.test.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import JobRole from '../model/jobRole.js';
+import JobRoleService from '../service/jobRoleService.js';
+import logger from '../service/logger.js';
+import { response } from 'express';
+
+const mockAxios = new MockAdapter(axios);
+const jobRoleValidatorStub = sinon.stub(new JobRoleValidator());
+
+const jobRoleTestEngi: JobRole = {
+    jobId: 69,
```
Comment (Fluffyguy611) [inline product-web-app/tests/jobRoleService.test.ts]: nice

File: product-web-app/views/add-new-job-role.html [hunk 11]
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
```
Comment (Fluffyguy611) [inline product-web-app/views/add-new-job-role.html]: Add parent.html, see main branch ./views/readme.md

**Review by Fluffyguy611 — APPROVED** [review]

File: September 2023.patch [hunk 12]
```diff
@@ -0,0 +1,1254 @@
+commit 0f57ca0effdf849c9bb0086d8623846192b8439a
```
Comment (ljercha) [inline September 2023.patch]: Remove that

File: product-backend/pom.xml [hunk 13]
```diff
@@ -33,6 +33,11 @@
     </dependencyManagement>
 
     <dependencies>
+        <dependency>
```
Comment (ljercha) [inline product-backend/pom.xml]: Remove all the changes from the pom they shouldn't be required

File: product-web-app/views/add-new-job-role.html [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+            </div>
+            {% endif %}
+            <form action="/admin/add-job-roles" method="post">
+                <div class="form-group">
+                    <label for="name">Name</label>
+                    <input type="text" name="name" class="form-control" value="{{ jobRole.name }}" required>
+                </div>
+                <div class="form-group">
+                    <label for="description">Description</label>
+                    <input type="text" name="description" value="{{ jobRole.description }}" class="form-control"
```
Comment (ljercha) [inline product-web-app/views/add-new-job-role.html]: Can you change in to the <textarea> so you can see the input in a cleaner way

File: product-backend/src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    private JobRoleDao jobRoleDao;
+    private JobRoleValidator jobRoleValidator;
+
+    public JobRoleService(JobRoleDao jobRoleDao, JobRoleValidator jobRoleValidator) {
+        this.jobRoleDao = jobRoleDao;
+        this.jobRoleValidator = jobRoleValidator;
+    }
+
+    public JobRoleService(JobRoleDao jobRoleDao, DatabaseConnector databaseConnector) {
+    }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/JobRoleService.java]: Remove
```suggestion
```

File: product-backend/src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+    public JobRole createNewJobRole(JobRoleRequest jobRole) throws FailedToCreateNewJobRoleException, SQLException, JobRoleAlreadyExistsException {
+
+        Optional<JobRole> optionalJobRole = jobRoleDao.createNewJobRole(jobRole);
+
+        if (optionalJobRole.isEmpty()) {
+            throw new FailedToCreateNewJobRoleException();
+        } else {
+            throw new JobRoleAlreadyExistsException();
+        }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/JobRoleService.java]: This code is always throwing an exception

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 17]
```diff
@@ -1,29 +1,47 @@
...
+import org.kainos.ea.model.JobRoleRequest;
+import org.kainos.ea.service.JobRoleService;
+import org.kainos.ea.service.JobRoleValidator;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-@Tag(name = "Job Roles")
+import java.sql.SQLException;
+
+@Tag(name = "Group A")
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java]: Job Role/ Job Roles

**Review by ljercha — APPROVED** [review]

**Pull request merged**